### ISSUE TITLE
BachGAN: High-Resolution Image Synthesis from Salient Object Layout

### ISSUE THREAD

**Issue by chaerinkong** [reyllama/paper-reviews#117]
*CVPR 2020*

# TL;DR

- Given salient layout (=foreground object layouts), retrieve most relevant ***semantic map*** from the ***memory bank***, fuse them through several ***conv layers*** and synthesize photo-realistic images out of them using ***SPADE***.

# Motivation

- Obtaining large semantic segmentation maps is difficult.
- Current L2I models (*namely layout2im, CVPR'19*) work only for low resolution and require both foreground and background layout.

# Previous Work

- Semantic Segmentation Map based models (ex: SPADE)
- Layout-to-Image models (ex: layout2im)

# Method

![image](https://user-images.githubusercontent.com/46308407/130962002-7605426b-f6d3-496f-840a-e2484b77f758.png)

## Background Retrieval Module
- Similar to IoU, compute ***layout similarity score*** and retrieve *m* most similar (Image, Semantic Map) pairs.
![image](https://user-images.githubusercontent.com/46308407/130962376-44caddaa-6595-4944-a125-b077b7969730.png)

## Background Fusion Module
- Take m [foreground label map ; background label map] feature maps and concatenate them to form a huge feature map.
![image](https://user-images.githubusercontent.com/46308407/130962620-52d9e735-bce2-452b-9534-df843894183b.png)
- Note that foreground label map is repeated m times in between every retrieved label map.
- Apply convolution with average pooling several times.
![image](https://user-images.githubusercontent.com/46308407/130962759-58144057-9415-45e7-a82f-e401608b54d9.png)
![image](https://user-images.githubusercontent.com/46308407/130962770-c16162e9-d4dd-4470-a977-bd1126aec191.png)

## Training Objective
![image](https://user-images.githubusercontent.com/46308407/130962789-eb4e4e1f-b7d8-42a7-b05a-bce491dfadd6.png)

## Image Generator
- From fused foreground-background feature map, use ResBlock + SPADE generator to synthesize photo-realistic images.

# Experiments

## Datasets
- Cityscapes 
- ADE20K

## Metric
- Segmentation Accuracy
- FID

## Baseline
![image](https://user-images.githubusercontent.com/46308407/130963799-8c679701-6516-49e1-a86e-aade7617d67e.png)
![image](https://user-images.githubusercontent.com/46308407/130963828-a9ff4e63-64d9-4436-9046-654555ce52ed.png)
![image](https://user-images.githubusercontent.com/46308407/130963901-0b05d78d-b7fe-4860-ab06-8690da6ada2b.png)

## Ablation
![image](https://user-images.githubusercontent.com/46308407/130963936-da0883ce-8edb-435e-b9ae-dfda83c18a77.png)
![image](https://user-images.githubusercontent.com/46308407/130963963-fc86ad8a-05f9-4720-80f7-b0b4fb8586fd.png)
- BachGAN-r uses only 1 retrieved semantic map.

## Analysis
![image](https://user-images.githubusercontent.com/46308407/130964072-9da36ada-879f-48b1-87ef-6a3ee4d5cdd9.png)
![image](https://user-images.githubusercontent.com/46308407/130964101-b839395e-9177-49eb-ba6e-8b998f0d2c0d.png)
![image](https://user-images.githubusercontent.com/46308407/130964126-77fae06a-d53e-43ac-abc1-13a759e8be27.png)
![image](https://user-images.githubusercontent.com/46308407/130964143-1437dcc8-1664-4d2e-99f1-0ba14a27694c.png)

# Comments
- Interesting idea to ***retrieve*** similar segmentation map for fine quality high resolution image synthesis.
- Method is very simple and straightforward, yet quite effective.
- However, the idea of ***retrieving*** samples from memory bank can limit its creative capacity; it's a compromise, but I'm not sure if it's worth it.
- One of the main reasons is that their evaluation heavily relies on Cityscape benchmark, which in my view is not that diverse. There is road in the center as a vertical line, and (presumably) some cars are to the left and right. If we look at the top 3 retrieved semantic maps, they look quite alike.
- Fundamentally, the model doesn't learn meaning of each label (object) or their relationships. Rather, it just *remembers* the most similar looking semantic map, and then simply rely on SPADE, which has been well proven to be powerful some time ago.
- In practice, this idea can be efficient, as in many cases, we don't have to reinvent the wheel. It's not hard to store several thousand samples in the memory bank. If that significantly boosts output quality, one could say why not. *(Though I want to invent something more creative and well reasoning about the visual world)*
